services/frontend-service/src/ui/ActionsCart/CheckoutDialog.tsx

```diff
@@ -78,11 +75,75 @@ const ApplyButton = (props: { openNotification: (msg: string) => void; closeDial
     }
 };
 
-export const CheckoutCart = () => {
-    const [openNotify, setOpenNotify] = React.useState(false);
-    const [dialogOpen, setDialogOpen] = React.useState(false);
-    const [notifyMessage, setNotifyMessage] = React.useState('');
+const showTextField = (actions: BatchAction[]) => {
+    for (const act of actions) {
+        const t = GetActionDetails(act).type;
+        if (t === ActionTypes.CreateEnvironmentLock || t === ActionTypes.CreateApplicationLock) {
+            return true;
+        }
+    }
+    return false;
+};
+
+const updateMessage = (act: BatchAction, m: string): BatchAction => {
+    switch (act.action?.$case) {
+        case 'createEnvironmentLock':
+            return {
+                action: {
+                    ...act.action,
+                    createEnvironmentLock: {
+                        ...act.action.createEnvironmentLock,
+                        message: m,
+                    },
+                },
+            };
+        case 'createEnvironmentApplicationLock':
+            return {
+                action: {
+                    ...act.action,
+                    createEnvironmentApplicationLock: {
+                        ...act.action.createEnvironmentApplicationLock,
+                        message: m,
+                    },
+                },
+            };
+        default:
+            return act;
+    }
+};
+
+const CheckoutButton = (props: { openDialog: () => void; disabled: boolean; setLockMessage: (m: string) => void }) => {
+    const { openDialog, disabled, setLockMessage } = props;
     const { actions } = useContext(ActionsCartContext);
+    const updateInput = useCallback((e) => setLockMessage(e.target.value), [setLockMessage]);
+
+    return (
+        <>
+            {showTextField(actions) && (
+                <TextField
+                    label="Lock Message"
+                    variant="outlined"
+                    sx={{ m: 1 }}
+                    placeholder="default-lock"
+                    onChange={updateInput}
+                    className="actions-cart__lock-message"
+                />
```

Comment:
I'm sorry, I don't quite understand the question. This is the input that will change the `lockMessage` so it should have the `@latest` value for `lockMessage`. right? 

at the start the `lockMessage` is empty and it's showing the placeholder and after that the `lockMessage` isn't changed anywhere else. so I think It works, unless there is some corner case or something?

Reply:
Since this calls `setLockMessage`, I would have expect for it to use the value of `lockMessage` as its content. In my own code I always (?) passed in a pair of value and setter, when I needed to set something. You did not run into it not displaying the `lockMessage`, because `TextField` implicitly also has a local copy of it?

Reply:
Yes, I believe that's how MUI designed the `TextField`